### DISCUSSION TITLE
Adventure map, Tree of Knowledge, free level up window [Done]

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#6725] [Ideas]
I think that even when the tree of knowledge does not require resources, the player should still be able to refuse the hero's level up ("YES" and "NO" buttons should replace the "OKAY" button in the window).

In fheroes2:

![Tree of knowledge 001](https://user-images.githubusercontent.com/43583994/221425536-130f558e-f1c9-49dd-ad29-85c47533affc.png)

![Tree of knowledge 002](https://user-images.githubusercontent.com/43583994/221425540-005965f5-fb93-4d59-be3b-74e8ab787636.png)

![Tree of knowledge 003](https://user-images.githubusercontent.com/43583994/221425547-f10948aa-cece-4797-971d-92f7743c984f.png)

**Comment by oleg-derevenetz** [comment]
> I think that even when the tree of knowledge does not require resources, the player should still be able to refuse the hero's level up ("YES" and "NO" buttons should replace the "CANCEL" button in the window).

For what reason should a player refuse to level up a hero for free?

**Reply by zenseii** [reply]
The level up is a one time thing, and so wasting it on just 1000 exp for the next level up on a level 1 hero, instead of saving it until later levels which would net a bigger experience point gain, is a common tactic. The reason you might even visit the tree at such an early stage without the intent to level up, is to find out if the tree will give it for free or if it will cost so that you know this for later when you feel ready to visit it.

I can see how such a change slightly alters the rules of the game, so if someone argues that having a "Yes/No" option will remove the game of chance that you're playing by even risking to visit the tree I wouldn't object much.

It is sort of similar to the whole "reorganize your troops in the followers dialog"-discussion that we had.

I believe later games allows to choose "Yes/No", but if I recall correctly none of the Trees in the later gave level ups for free either.

**Reply by zenseii** [reply]
I'd like to note that I don't remember if in Heroes 2 the Tree of Knowledge give you whatever remains until your next level, e.g. if you need 73 to level up it will give you just that, or if it will always give you the amount needed from your current level to the next, e.g. if your hero is level 1 he'll get 1000 exp whether he has 0 exp or 999 exp. In later games it works the latter way.

**Reply by oleg-derevenetz** [reply]
If player wants to conserve the Tree of Knowledge for late game, then he will just do it regardless of cost. In late game 2000 gold or 10 gems are not any significant amounts. These amounts are only significant for the early game, and if a player wants his hero to visit the Tree of Knowledge at the beginning of the game, then he will most likely be happy with a free level up.

**Reply by LeHerosInconnu** [reply]
Hello @oleg-derevenetz and @zenseii,

> I'd like to note that I don't remember if in Heroes 2 the Tree of Knowledge give you whatever remains until your next level, e.g. if you need 73 to level up it will give you just that, or if it will always give you the amount needed from your current level to the next, e.g. if your hero is level 1 he'll get 1000 exp whether he has 0 exp or 999 exp. In later games it works the latter way.

In Heroes 2, it is the number of experience points to go from one level to the next that is given and not just the number of points needed.
In fheroes2, before, just the number of points needed was given, but this has since been corrected to work like in Heroes 2.

> If player wants to conserve the Tree of Knowledge for late game, then he will just do it regardless of cost. In late game 2000 gold or 10 gems are not any significant amounts. These amounts are only significant for the early game, and if a player wants his hero to visit the Tree of Knowledge at the beginning of the game, then he will most likely be happy with a free level up.

@zenseii explained it well.

And, if the player wants his hero to take the free level at the beginning, he will always be able to do so.

The modification just gives more choices, depending on the player's wishes and the way he plays, with more flexibility in how to do it, thus with a better game experience.

**Comment by ihhub** [comment]
This is micromanagement. If a player has too many choices it's bad. No player so far complained about these particular mechanics or asked to change it.

Now checking what we had in the past the trend was to reduce the micromanagement like automated collection of resources from Water Wheels. We should keep in mind that some players have no experience with the game at all. So me as a new player for instance visits this object and sees a window with yes and no for free stuff. Why you should even ask for free stuff? It's free, of course I will always accept it.

Before implementing I would like to gather the opinion of the rest of the team.

**Comment by LeHerosInconnu** [comment]
Hello @ihhub,

> This is micromanagement. If a player has too many choices it's bad. No player so far complained about these particular mechanics or asked to change it.
> 
> Now checking what we had in the past the trend was to reduce the micromanagement like automated collection of resources from Water Wheels. We should keep in mind that some players have no experience with the game at all. So me as a new player for instance visits this object and sees a window with yes and no for free stuff. Why you should even ask for free stuff? It's free, of course I will always accept it.
> 
> Before implementing I would like to gather the opinion of the rest of the team.

But this reduces the micromanagement of having another hero visit the trees of knowledge to optimize the advancement of the main hero (or heroes).
With the enhancement, the main hero can visit the tree of knowledge directly without risking to "burn out" an advancement.

**Comment by Branikolog** [comment]
Hi all!
I personally would like to make choice on my own, whether to level up my hero currently for free, or to come later to have more benefit from the experience quantity point. The only thing I'm concerning, that all three cases would have to be clearly distinctive. Players don't read much texts. Original implementation had either dialog window (yes/no) or simple window with the only button, which visually tells player, whether he need to pay something (catch his attention and makes him read the text to find the price for level up) or it's free, which does not requires anything more than accepting the bonus.
So, on implementing the proposed thing we need to rework the dialog window for the tree of knowledge... Probably adding gold or gem icons somewhere in that window.

**Reply by zenseii** [reply]
Hi everyone! I don't see this as adding more micromanagement, but rather removing it as stated by @LeHerosInconnu.

My only point against this is the one I explained above about how this potentially changes your decision-making about whether you should visit a tree of knowledge or not.

In the grander scheme this might lead to heroes having higher levels because players will have an easier time postponing their visit to the tree of knowledge. It is debatable how often such a situation will happen compared to before this change, because usually these trees are scarse and not always located right next to your base.

There is also the small argument that players from later games will not be frustrated by being given the level up right away, unlike how they are used to always having a choice in the later games. However, this is something any player quickly adjusts to imo.

Myself, while playing fheroes2 I always visit a tree with my scout/courier heroes or those that will see little battle and await going there with my champion, unless I know I will never come back to that place again - although later during a game when you have the town portal and/or dimension door spells you'll be able to return to most places on the map quite easily.

**Comment by LeHerosInconnu** [comment]
This has been implemented.
I close the discussion.
